### PR TITLE
convert divcoeff.c to D

### PR DESCRIPTION
**Pull request by WalterBright**
first one in the back end!

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
There's a Visual Stdio project file as well.

**Comment by WalterBright** [comment]
I'm having enough trouble getting these makefiles to work.

**Comment by WalterBright** [comment]
It looks like circleci is using an older dmd to compile. It needs to be upgraded to 2.074.1

**Comment by wilzbach** [comment]
> It looks like circleci is using an older dmd to compile. It needs to be upgraded to 2.074.1

Simply bump it here:

https://github.com/dlang/dmd/blob/master/circleci.sh#L5

**Comment by WalterBright** [comment]
Thanks! That did the trick on circleci.

**Comment by MartinNowak** [comment]
As said per mail, neither gdc nor ldc support anything like `-mv` which would restrict us to dmd to build releases. Looking at ~30% speedup for dmd using a better optimizing compiler, that doesn't seem like a smart move.

**Comment by MartinNowak** [comment]
Sorry, doesn't look like this is using `-mv` at all. For what would that be needed?

**Comment by MartinNowak** [comment]
Some info on why you need to update the required compiler would be helpful!
Also need to bump the auto_bootstrap version in posix.mak.

File: src/ddmd/backend/divcoeff.d [hunk 1]
```diff
@@ -6,35 +6,44 @@
  * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
  * License:     Distributed under the Boost Software License, Version 1.0.
  *              http://www.boost.org/LICENSE_1_0.txt
- * Source:      https://github.com/dlang/dmd/blob/master/src/ddmd/backend/divcoeff.c
+ * Source:      https://github.com/dlang/dmd/blob/master/src/ddmd/backend/divcoeff.d
```
Comment (wilzbach) [inline src/ddmd/backend/divcoeff.d]: How about?

```
$(DMDSRC backend/_divcoeff.d)
```

**Comment by ibuclaw** [comment]
My guess is that some compilers (or ancient perl wrappers) don't understand `-betterC`.

**Comment by MartinNowak** [comment]
> My guess is that some compilers (or ancient perl wrappers) don't understand -betterC.

Ah yes, [that](https://travis-ci.org/dlang/dmd/jobs/243205245#L588)'s another problem. Why do we need `-betterC` for the backend? Is this some `dmc` tribute?

**Comment by ibuclaw** [comment]
> Ah yes, that's another problem. Why do we need -betterC for the backend? Is this some dmc tribute?

Well, it looks like `-betterC` is the reason why jenkins fails.

https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fdmd/detail/PR-6907/9/pipeline#step-74-log-128

You *could* workaround by building with `-release`, but I already am not a fan of the travisci tests doing that.

**Comment by WalterBright** [comment]
> Some info on why you need to update the required compiler would be helpful!

The older compiler generated code that was looking for the module assert() function in `-betterC` mode, which had been removed.

**Comment by WalterBright** [comment]
Essentially, it's time to move forward with converting the backend to D. To that end, gdc, ldc, and the build process needs to support:
1. -mv switch
2. removal of the generated assert helper functions
3. https://github.com/dlang/dmd/pull/6901
None of these are tricky or difficult. To that end, divcoeff.d is a trial run to see where the issues are in the toolchain for doing this. Looks like it's doing its job :-)

**Comment by ibuclaw** [comment]
> #6901

Making compilers support features that have not made it to stable (or even in a release) is probably asking for a bit too much.

**Comment by WalterBright** [comment]
> Making compilers support features that have not made it to stable (or even in a release) is probably asking for a bit too much.

These changes are simple and not controversial. Let's move forward with them.

File: src/ddmd/backend/divcoeff.d (hunk 1 above)
Comment (andralex) [inline src/ddmd/backend/divcoeff.d]: yes please

**Comment by andralex** [comment]
OK I'll pull the switch on this to get in once it passes tests.

**Review by andralex — APPROVED** [review]

**Comment by wilzbach** [comment]
CC @CyberShadow for updating DAutoTest

**Comment by CyberShadow** [comment]
@WalterBright Just so we're on the same page, please describe the procedure for building DMD master (incl. this pull request) from scratch without a host DMD compiler (C++ only).

**Comment by MartinNowak** [comment]
Didn't we just agree that 30% speedup from building dmd with gdc is worthwhile for an upcoming release? This would block this option for quite a while. Also jumping ahead with using latest edge features sets quite a bad precedence for dmd/gdc/ldc compatibility.
But most important, it's still unclear why we even need `-betterC` here.

**Comment by ibuclaw** [comment]
Consider that the oldest compiler building DMD IIRC is using v2.066 of the front-end.  I think it's safe to do a leap to 2.068, possibly 2.070 at latest, but I don't really want to break builds when using a release that is not strictly that old.

**Comment by WalterBright** [comment]
> please describe the procedure for building DMD master (incl. this pull request) from scratch without a host DMD compiler (C++ only).

That's a great question, but since building DMD already requires a D compiler, it is off-topic for this PR. Let's keep this thread focused on this PR.

**Comment by WalterBright** [comment]
> Didn't we just agree that 30% speedup from building dmd with gdc is worthwhile for an upcoming release?

Yes.

> This would block this option for quite a while.

Not really. The 3 changes needed to the D compiler to support this, as described, are easy and minimal, with little risk of disruption.

> Also jumping ahead with using latest edge features sets quite a bad precedence for dmd/gdc/ldc compatibility.

Getting DMD in 100% D also has advantages. I've been waiting since the front end was converted to D to get moving on this. How many years should this take?

> Most important though, it's still unclear why we even need -betterC here.

1. -betterC needs to move forward. It'll never happen until we start dogfooding it.

2. The back end is designed to run with asserts on. Turning them off, like is done in the front end, is unacceptable. If -release were left off, the invariant checks for struct member function calls is also unacceptable. -betterC makes the asserts behave like C asserts.

**Comment by CyberShadow** [comment]
> it is off-topic for this PR

It's about as on-topic as updating CI services to support this change, which is not going to happen without discussion and acknowledgement from all sides of the full impact of changes like these.

**Comment by ibuclaw** [comment]
> Just so we're on the same page, please describe the procedure for building DMD master (incl. this pull request) from scratch without a host DMD compiler (C++ only).

I wouldn't use anything other than a release compiler from the downloads page for DAutoTest.  We have bootstrap tests as part of the travisCI scripts.

**Comment by ibuclaw** [comment]
> Not really. The 3 changes needed to the D compiler to support this, as described, are easy and minimal, with little risk of disruption.

It would be a major disruption, as suddenly you cannot use the current DMD release, nor any of the prior eight major releases to build master.  Effectively our entire CI infrastructure will be broken.

However, I'm all up for trimming the hedges a bit, if for @CyberShadow that means supporting a D compiler that can be built using C++, then we can set the new baseline at 2.068 - dropping 2.066 and 2.067 (ping @braddr for confirmation that auto-tester is using >= 2.068).

If dropping the DMD C++ bootstrap requirement altogether is more desired, I would be OK to stretch up until 2.071, which is approximately the feature set supported by GDC (no binaries available though).

You're not the first person to try submitting a patch that depends on a feature more recent than 2.066, and as a rule of thumb, there really isn't that much of value in more recent versions of DMD that would be a showstopper for anything that we'd want included in the compiler implementation.

**Comment by CyberShadow** [comment]
> supporting a D compiler that can be built using C++, then we can set the new baseline at 2.068

IIRC, 2.068 is the first version that can no longer be (directly) built with just a C++ compiler.

> If dropping the DMD C++ bootstrap requirement altogether is more desired,

I'm guessing you actually mean adding another step to the bootstrap process?

**Comment by ibuclaw** [comment]
> IIRC, 2.068 is the first version that can no longer be (directly) built with just a C++ compiler.

Ah, I'm forgetting about idgen.d - I was only thinking about the rest of the compiler.

> I'm guessing you actually mean adding another step to the bootstrap process?

I meant using a front end that is entirely implemented in D.

**Comment by wilzbach** [comment]
@WalterBright regarding:

>> Didn't we just agree that 30% speedup from building dmd with gdc is worthwhile for an upcoming release? This would block this option for quite a while. Also jumping ahead with using latest edge features sets quite a bad precedence for dmd/gdc/ldc compatibility.
>> Most important though, it's still unclear why we even need -betterC here.


> Not really. The 3 changes needed to the D compiler to support this, as described, are easy and minimal, with little risk of disruption.

Hehe, it seems that reality proves that it's not minimal.

> -betterC needs to move forward. It'll never happen until we start dogfooding it.

It would also help `-betterC`, if we document it for a start: https://github.com/dlang/dlang.org/pull/1796
In any case, as far as I can see the situation you have three options:

(A) Wait and hope that LDC/GDC incorporate your changes soon (for LDC that might happen fairly soon, but idk about GDC)
(B) Accept [DIP1006](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1006.md) and build the backend without `-betterC`, but with `assert`s
(C) Roll your own runtime `assert`

> The back end is designed to run with asserts on. Turning them off, like is done in the front end, is unacceptable. If -release were left off, the invariant checks for struct member function calls is also unacceptable. -betterC makes the asserts behave like C asserts.

Then accept [DIP1006](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1006.md) ;-)

**Comment by kinke** [comment]
> (A) Wait and hope that LDC/GDC incorporate your changes soon (for LDC that might happen fairly soon, but idk about GDC)

LDC 1.3, released a week ago, supports `-mv`.

**Comment by WalterBright** [comment]
@kinke thanks, that's great news! add betterC too, please!

**Comment by kinke** [comment]
We also support `-betterC`, but I think it only skips the emission of ModuleInfos. I hope that suffices. ;)

**Comment by WalterBright** [comment]
@wilzbach I approved https://github.com/dlang/dlang.org/pull/1796 pending a couple trivial edits.

**Comment by MartinNowak** [comment]
We could just add `-betterC` as noop to the gdmd wrapper script.
https://github.com/D-Programming-GDC/GDMD/blob/master/dmd-script

**Comment by andralex** [comment]
OK folks, status of this? We need to plow forward. Thanks! cc @ibuclaw @MartinNowak

**Comment by kinke** [comment]
I implemented `-betterC` for LDC in https://github.com/ldc-developers/ldc/pull/2365. We've just released LDC 1.5-beta1, so I'd guess that one more month would be needed for 1.6-beta1 which would include that PR.

**Comment by WalterBright** [comment]
> one more month would be needed

Great! Ready when you are! @ibuclaw ?

**Comment by andralex** [comment]
ping @ibuclaw

**Comment by ibuclaw** [comment]
Wouldn't want to give a precise date, but gcc-8 is first this would land in.  As @MartinNowak said, you are free to make `-betterC` a noop for the dmd-script wrapper.

**Comment by WalterBright** [comment]
> gcc-8 is first this would land in

When is that expected?

**Comment by joakim-noah** [comment]
[Looks like early next year](https://www.phoronix.com/scan.php?page=news_item&px=GCC-8-Stage-3-Soon).

**Comment by wilzbach** [comment]
Regarding the current failure: I think you need to add `G_DOBJS` to the new `unittest` target in the `posix.mak` or probably better: add it to `D_SRCS`.

**Comment by JackStouffer** [comment]
Walter, the bot comment is going to keep coming back every time there's a new commit. If it doesn't see it's own comment to update, it makes a new one.

**Comment by WalterBright** [comment]
> If it doesn't see it's own comment to update, it makes a new one.

That's not working, then, because I'd delete 10 of them in a row, not one at a time.

**Comment by andralex** [comment]
so... pull?

**Comment by wilzbach** [comment]
> so... pull?

AFAICT we are still waiting for GDC to catch up or someone to fork the `gdmd` script and make `-betterC` a noop:

> Wouldn't want to give a precise date, but gcc-8 is first this would land in. As @MartinNowak said, you are free to make -betterC a noop for the dmd-script wrapper.

**Comment by WalterBright** [comment]
Rebooted as https://github.com/dlang/dmd/pull/7714 because I could not get `git rebase` to work with renamed files (it kept insisting I had not merged them).

**Comment by wilzbach** [comment]
> I could not get git rebase to work with renamed files (it kept insisting I had not merged them).

Not sure what you did, but it works fine for me:

```shell
> go pr/6907
> git rebase upstream/master
First, rewinding head to replay your work on top of it...
Applying: convert divcoeff.c to D
Using index info to reconstruct a base tree...
A	circleci.sh
A	src/ddmd/backend/cod2.c
A	src/ddmd/backend/divcoeff.c
M	src/posix.mak
M	src/win32.mak
Falling back to patching base and 3-way merge...
Auto-merging src/win32.mak
Auto-merging src/posix.mak
CONFLICT (content): Merge conflict in src/posix.mak
CONFLICT (rename/rename): Rename "src/ddmd/backend/divcoeff.c"->"src/dmd/backend/divcoeff.c" in branch "HEAD" rename "src/ddmd/backend/divcoeff.c"->"src/ddmd/backend/divcoeff.d" in "convert divcoeff.c to D"
Auto-merging src/dmd/backend/cod2.c
Auto-merging .circleci/run.sh
error: Failed to merge in the changes.
Patch failed at 0001 convert divcoeff.c to D
The copy of the patch that failed is found in: .git/rebase-apply/patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

> git mergetool

Merging:
src/ddmd/backend/divcoeff.c
src/ddmd/backend/divcoeff.d
src/dmd/backend/divcoeff.c
src/posix.mak

Deleted merge conflict for 'src/ddmd/backend/divcoeff.c':
  {local}: deleted
  {remote}: deleted
Use (m)odified or (d)eleted file, or (a)bort? d

Deleted merge conflict for 'src/ddmd/backend/divcoeff.d':
  {local}: deleted
  {remote}: created file
Use (c)reated or (d)eleted file, or (a)bort? c

Deleted merge conflict for 'src/dmd/backend/divcoeff.c':
  {local}: created file
  {remote}: deleted
Use (c)reated or (d)eleted file, or (a)bort? d

Normal merge conflict for 'src/posix.mak':
  {local}: modified file
  {remote}: modified file
Hit return to start merge resolution tool (kdiff3): 
```

**Pull request closed without merging**